src/Services/Air/Air.js

```diff
@@ -409,16 +409,13 @@ module.exports = (settings) => {
           }
         ));
       };
-
+     
       return this.getUniversalRecordByPNR(options)
-        .then((ur) => {
-          const urr = Array.isArray(ur) ? ur[0] : ur;
-          const record = {
-            reservationLocatorCode: urr.uapi_reservation_locator
-          };
-          return (ignoreTickets
-            ? Promise.resolve([])
-            : this.getTickets(record).then(checkTickets)
+        .then(() => {
+          return (
+            ignoreTickets
+              ? Promise.resolve([])
+              : this.getTickets(options).then(checkTickets)
```

Comment:
Hello Ukeh, thank you a lot for your participation!

I checked the getTickets method of the Air class and thought that  we can make a bit different update:
`this.getTickets(record).then(checkTickets)` -> `service.getTickets(record).then(checkTickets)`
What do you think?

Thank you, and have a nice day!

Reply:
this replaced by service, both does the same job though

Reply:
What do you think?

Thanks for your review.

Reply:
Hi @ukeh!
Thank you for the contribution.

You can see, that `service` is created on line 12 and is returned from `AirService.js`. It has a bit different set of methods and provides some lower level operations. For instance `getTickets` of `service` accepts `reservationLocatorCode`, not `pnr` as you can see when it called on line 281.

You can also see, that on line 275 method `this.getTickets` calls for `this.getUniversalRecordByPNR` which is already called in method `cancelBooking` you edit, and it is actually called to receive `reservationLocatorCode`.

So simply replacing `this.getTickets` to `service.getTickets` will do the trick.

I'd also ask you not to modify other lines of code (like the line I mentioned before and the other one with condition), so the changeset will have smaller footprint.

Once again, thanks for finding this one!

Reply:
There is also another way to make it even more nice: it is to call `this.getTickets` instead of `this.getUniversalRecordByPNR` and then `service.getTickets`.

The code should look like:

```javascript
    cancelBooking(options) {

      /* Other method stuff */

      return (
        ignoreTickets
          ? Promise.resolve()
          : this.getTickets(options).then(checkTickets)
      )
        .then(() => this.getBooking(options))
        .then(booking => service.cancelBooking(booking))
        .catch(
          err => Promise.reject(new AirRuntimeError.FailedToCancelPnr(options, err))
        );
    }
```

Reply:
in have tried this as u can see in my thrid commit
but when am doing this the test case is failing , so i reverted that

Reply:
> 

I also noted the point that u have mentioned. The get tickets function actually needs reservationLocatorCode, the reason I never updated the function is that I thought it might be used somewhere else, and changing it may cause some troubles in other functionalities.

Reply:
cancelBooking(options) {
      const ignoreTickets = typeof options.ignoreTickets === 'undefined'
        ? false // default value
        : options.ignoreTickets;

      const checkTickets = (tickets) => {
      
      return this.getUniversalRecordByPNR(options)
        .then((ur) => {
          const urr = Array.isArray(ur) ? ur[0] : ur;
          const record = {
            reservationLocatorCode: urr.uapi_reservation_locator
          };
          return (ignoreTickets
            ? Promise.resolve([])
            : service.getTickets(record).then(checkTickets)
          )
            .then(() => this.getBooking(options))
            .then(booking => service.cancelBooking(booking))
            .catch(
              err => Promise.reject(new AirRuntimeError.FailedToCancelPnr(options, err))
            );
        });
    },



Tried as per ur suggestion but around 8 test cases getting failed. 


8 failing

  1) #AirService
       cancelBooking
         should cancel PNR if no tickets available:
     expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
  AssertionError: expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      at /home/unnikrishnan/Desktop/uapi-edit/uapi-json/test/Air/Air.test.js:1395:51

  2) #AirService
       cancelBooking
         should cancel PNR if tickets have only VOID coupons:
     expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
  AssertionError: expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      at /home/unnikrishnan/Desktop/uapi-edit/uapi-json/test/Air/Air.test.js:1429:51

  3) #AirService
       cancelBooking
         should fail with AirRuntimeError.PNRHasOpenTickets PNR if tickets have OPEN coupons and no cancelTicket option:
     expected getUniversalRecordByPNR to have been called exactly twice, but it was called once
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
  AssertionError: expected getUniversalRecordByPNR to have been called exactly twice, but it was called once
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      at /home/unnikrishnan/Desktop/uapi-edit/uapi-json/test/Air/Air.test.js:1478:51

  4) #AirService
       cancelBooking
         should succeed when there are no VOID, but REFUNDED tickets:
     expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
  AssertionError: expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      at /home/unnikrishnan/Desktop/uapi-edit/uapi-json/test/Air/Air.test.js:1517:51

  5) #AirService
       cancelBooking
         should succeed when there are VOID and REFUNDED tickets:
     expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
    getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
  AssertionError: expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      getUniversalRecordByPNR({ pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      at /home/unnikrishnan/Desktop/uapi-edit/uapi-json/test/Air/Air.test.js:1565:51

  6) #AirService
       cancelBooking
         should succeed when there are OPEN and VOID tickets and cancelTickets = true:
     expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
    getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
    getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
  AssertionError: expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
      getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      at /home/unnikrishnan/Desktop/uapi-edit/uapi-json/test/Air/Air.test.js:1620:51

  7) #AirService
       cancelBooking
         should fail with AirRuntimeError.PNRHasOpenTickets PNR if tickets have coupons other than OPEN:
     expected getUniversalRecordByPNR to have been called exactly twice, but it was called once
    getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
  AssertionError: expected getUniversalRecordByPNR to have been called exactly twice, but it was called once
      getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      at /home/unnikrishnan/Desktop/uapi-edit/uapi-json/test/Air/Air.test.js:1671:51

  8) #AirService
       cancelBooking
         should cancel tickets and PNR if no errors occured:
     expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
    getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
    getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (/home/unnikrishnan/Desktop/uapi-edit/uapi-json/src/Services/Air/Air.js:130:30)
  AssertionError: expected getUniversalRecordByPNR to have been called exactly thrice, but it was called twice
      getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      getUniversalRecordByPNR({ cancelTickets: true, pnr: "PNR001", viewOnly: false }) => [Promise] {  } at Object.getUniversalRecordByPNR (src/Services/Air/Air.js:130:30)
      at /home/unnikrishnan/Desktop/uapi-edit/uapi-json/test/Air/Air.test.js:1720:51

Reply:
Test cases are failing because the structure of calls is changing, and tests should be updated accordingly.

I'll do update myself according to the issue you've created #589 

Thank you Ince again!

Reply:
Thank you and hope the issue get fixed and released soon.